### ISSUE TITLE
Runtime change Subtree path.

### ISSUE THREAD

**Issue by ODtian** [MBoqui/event-driven-behavior-tree#7]
Is there a way to change subtree path dynamicly during runtime?

**Comment by MBoqui** [comment]
Not easily.
You need a reference to the node to change its path and you need to reset the tree somehow, since the path is only read on `_setup_execution`. If you don't care for waiting until the tree resets itself, you only need the reference to the path. Also, this will set the path for that tree resource, so all agents using it will be changed at once.

The implementation I have planned is:
- Register the subtree node to a blackboard key on initialization)
- When checking monitor interruption for subtree, if path changed, interrupt at this node

This way each agent can have its own value for that subtree and each instance of the subtree node for that agent can have the same or a different tree assigned.

**Comment by ODtian** [comment]
I have some questions about how to write game logic with subtree

- How to decide which bt to execute? For top level it's my npc, however what about subtree's subtree? I can't put all logic in the a same npc bt agent, it's hard to reuse and organize them.
- Can subtree have it's own logic and - methods? it's basically question1 but more common. A common scene is that npc use item, and for convenience item provide it behavior as a subtree to npc to execute, and item's behavior might require it's own logic. Subtree's logic is unknown to top level agent.

Maybe i was trapped somewhere, what's missing?

**Comment by MBoqui** [comment]
There are two ways to add logic to a tree:
- Extend BTAgent and add your function there, then call it using a BTCallable node. This is simpler but can overcrowd the agent script.
- Create a custom node that extends BTLeaf with your code. This needs more work to create the node but is more modular, since you can use the nodes only on trees that require them.

Currently the subtree is more useful as a tool to avoid repeating code/tree structure than to inject new behavior in a tree. I use it for repetitive actions, like moving. I set a blackboard key `"move_target"` and call my move subtree. The subtree deals with everything from pathfinding, to speed calculation based on the context (checking some blackboard keys and setting others), and dealing with opening/closing doors in the path.

For dealing with items, you could create a new agent for your item containing its specific behavior and in your NPC tree call a function or custom node to use that item. You only need to set your item to a blackboard key like `"equipped_item"`:
``` GDScript
#item_agent.gd
extends BTAgent
var _result : BTItem.BTState
var _npc_agent 

# this will store the result returned by the top node to the agent
func _on_tree_result_reported(agent : BTAgent, result : BTItem.BTState) -> void:
    super._on_tree_result_reported(agent, result)
    _result = result

func use(npc_agent) -> BTItem.BTState:
    _npc_agent = npc_agent # so you can access the npc from the item tree, maybe store in blackboard
    tick()
    return _result # forward result from item tree back to npc tree

# you can extend this script to add code to individual items without having to create several custom nodes and not overcrowding one agent script.

#npc_agent.gd
extends BTAgent

func use_equipped_item() -> BTItem.BTState:
    var equipped_item = blackboard.get_value("equipped_item")
    if not equipped_item: return BTItem.BTState.FAILURE

    return equipped_item.use(self)
```

This is basically an implementation of a subtree using agents. I have not tested this, but i think It should work. If you do test it, could you report how it worked for you? It might be interesting to create a proper workflow for this, since its even more flexible than my planned subtree implementation. Even if it has the cost of needing another agent and blackboard.

**Comment by ODtian** [comment]
Yes, it works perfectly for my simple case, I think it should be in workflow. And there are some details issues:
- Running subtree require an agent instance, so one item agent can't run in multiple npc agent at same time, should we just dupulicate it?
- Calling npc agent method in item method can be supported by `parent_agent.method` directly filling method name, it saves a lot of time.

**Comment by MBoqui** [comment]
Each agent will store its own internal state inside the blackboard, so one instance per object should be fine, but you will need to handle this state inside the item tree. 

Imagine two NPCs trying to use a microwave oven to heat their food in a Sims-like game:
- A uses the microwave and it returns `RUNNING` setting its `"food"` entry in the blackboard to `CupNoodles` instance in the player inventory and `"timer"` to 60 seconds.
- I don't think the microwave should wait for A to keep ticking it. Each object should handle its own logic. The microwave needs to tick itself to pass the time while A waits. This means it will need to differentiate between a call from another agent and one from itself inside the tree, this can be a custom condition that checks if the `parent_agent` is set (important to clear it somewhere). This also frees the NPC to do other tasks while they wait for the microwave to return not running.
- now B wants to heat their `Pasta` in the microwave. By checking the state of the microwave we can have B:
  - Wait for A to pick up their food.
  - Eat A's food.
  - Throw the CupNoodles in the trash and put the Pasta inside.

This way A and B can have different behaviors upon using the microwave, based on their individual characteristics and the state of the item. If A dies while the food is in the microwave, it might finish heating, someone puts it in a table nearby and then the item is free to use again. All contained inside the microwave tree.

I don't know your use case, maybe adding a new agent for the microwave for each NPC interacting with it is a good option. If implementing this chaining agents behavior for the plugin I would make something simple like the `item_agent.gd` from my previous comment (just forwarding result, tick call and having a reference to the caller_agent) and leave the details for the developer, since the implementation should vary with the use cases.

**Comment by ODtian** [comment]
The microwave example makes a lot of sense, and one bt instance per agent node should be enough for common scene which not requiring high perfermence. 

Maybe next step is implementing features:
- assign subtree path in blackboard
- assign an agent as a subtree
